File: src/price_model.py
import datetime
import pickle
from typing import Dict, List
from collections import deque
from exchangebookgenerator import ExchangeBookGenerator
from modelisation.simulate_quotes import QuoteGenerator

DEFAULT_SHARES = ExchangeBookGenerator.DEFAULT_SHARES.copy()
DEFAULT_OFFSETS = ExchangeBookGenerator.DEFAULT_OFFSETS.copy()

class PriceModel:
    """Génère et met à jour les prix du carnet tout en détectant les pivots
    au moyen d’un buffer rétrospectif de 2*window+1 secondes."""

    def __init__(
        self,
        model_path: str = "../models/model-20251029-100049.pkl",
        tick_size: float = 0.01,
        spread: float = 0.02,
        window_seconds: int = 30,
    ) -> None:
        # --- modèle de génération des ticks -----------------------------------
        with open(model_path, "rb") as f:
            quote_model = pickle.load(f)

        base_bid = 100.00
        base_ask = base_bid + spread

        self.generator = QuoteGenerator(quote_model, base_bid, base_ask)
        self.exchange_gen = ExchangeBookGenerator(
            shares=DEFAULT_SHARES,
            offsets_p=DEFAULT_OFFSETS,
            seed=42,
        )
        self.last_exchange_quotes: List[Dict] = []

        # --- paramètres --------------------------------------------------------
        self.tick_size = tick_size
        self.spread = spread
        self.window = window_seconds

        # --- buffer rétrospectif : longueur = 2*window + 1 --------------------
        self.buffer: deque[Dict[str, float]] = deque(
            maxlen=2 * window_seconds + 1
        )
        for _ in range(self.buffer.maxlen):
            self.buffer.append(
                {"bid": base_bid, "ask": base_ask, "mid": (base_bid + base_ask) / 2}
            )

        # --- état pivot & best bid/ask ----------------------------------------
        self.pivot: str | None = None          # "PL", "PH" ou None
        self.pivot_bid = base_bid
        self.pivot_ask = base_ask
        self.best_bid = base_bid
        self.best_ask = base_ask

    # -------------------------------------------------------------------------
    def update(self) -> None:
        """Ajoute le tick courant, met à jour le buffer et détecte
        un éventuel pivot situé window s dans le passé."""
        now_dt = datetime.datetime.now(datetime.timezone.utc)

        # 1) Génération du tick à l’instant présent (plus de look-ahead)
        ticks = self.generator.step_second(now_dt, ("N", "O"), "F")
        if ticks:                                  # un vrai tick a été produit
            last = ticks[-1]
            self.last_exchange_quotes = self.exchange_gen.generate(last)
            bid = last["priceBid"]
            ask = last["priceAsk"]
            mid = (bid + ask) / 2
            self.best_bid, self.best_ask = bid, ask
        else:                                      # aucune variation → reprise du dernier prix
            last_entry = self.buffer[-1]
            bid, ask, mid = (
                last_entry["bid"],
                last_entry["ask"],
                last_entry["mid"],
            )

        # 2) Mise à jour du buffer (append à droite, pop automatique à gauche)
        self.buffer.append({"bid": bid, "ask": ask, "mid": mid})

        # 3) Détection du pivot sur l’élément situé window s dans le passé
        if len(self.buffer) == self.buffer.maxlen:
            centre = self.buffer[self.window]
            mids = [e["mid"] for e in self.buffer]

            if centre["mid"] == max(mids):
                self.pivot = "PH"                 # pivot high
            elif centre["mid"] == min(mids):
                self.pivot = "PL"                 # pivot low
            else:
                self.pivot = None

            self.pivot_bid = centre["bid"]
            self.pivot_ask = centre["ask"]

File: src/models/PriceModel.ts
// PriceModel.ts - Converted from price_model.py

import { QuoteModel, QuoteGenerator } from "./QuoteModel";
import { ExchangeBookGenerator, Quote } from "./ExchangeBookGenerator";

interface BufferEntry {
  bid: number;
  ask: number;
  mid: number;
}

export class PriceModel {
  private generator: QuoteGenerator;
  private exchangeGen: ExchangeBookGenerator;
  private window: number;
  private buffer: BufferEntry[];
  private maxBufferSize: number;

  public lastExchangeQuotes: Quote[];
  public pivot: "PL" | "PH" | null;
  public pivotBid: number;
  public pivotAsk: number;
  public bestBid: number;
  public bestAsk: number;

  constructor(
    quoteModel: QuoteModel,
    _tickSize: number = 0.01,
    spread: number = 0.02,
    windowSeconds: number = 30
  ) {
    const baseBid = 100.0;
    const baseAsk = baseBid + spread;

    this.generator = new QuoteGenerator(quoteModel, baseBid, baseAsk);
    this.exchangeGen = new ExchangeBookGenerator(undefined, undefined, 42);
    this.lastExchangeQuotes = [];

    this.window = windowSeconds;

    // Initialize buffer
    this.maxBufferSize = 2 * windowSeconds + 1;
    this.buffer = [];
    for (let i = 0; i < this.maxBufferSize; i++) {
      this.buffer.push({
        bid: baseBid,
        ask: baseAsk,
        mid: (baseBid + baseAsk) / 2,
      });
    }

    this.pivot = null;
    this.pivotBid = baseBid;
    this.pivotAsk = baseAsk;
    this.bestBid = baseBid;
    this.bestAsk = baseAsk;
  }

  update(): void {
    const nowDt = new Date();

    // Generate tick
    const ticks = this.generator.stepSecond(nowDt, ["N", "O"], "F");

    let bid: number;
    let ask: number;
    let mid: number;

    if (ticks.length > 0) {
      const last = ticks[ticks.length - 1];
      this.lastExchangeQuotes = this.exchangeGen.generate(last);
      bid = last.priceBid;
      ask = last.priceAsk;
      mid = (bid + ask) / 2;
      this.bestBid = bid;
      this.bestAsk = ask;
    } else {
      // No variation, reuse last price
      const lastEntry = this.buffer[this.buffer.length - 1];
      bid = lastEntry.bid;
      ask = lastEntry.ask;
      mid = lastEntry.mid;
    }

    // Update buffer (append to end, remove from beginning)
    this.buffer.push({ bid, ask, mid });
    if (this.buffer.length > this.maxBufferSize) {
      this.buffer.shift();
    }

    // Detect pivot at window position in the past
    if (this.buffer.length === this.maxBufferSize) {
      const centre = this.buffer[this.window];
      const mids = this.buffer.map((e) => e.mid);

      const maxMid = Math.max(...mids);
      const minMid = Math.min(...mids);

      if (centre.mid === maxMid) {
        this.pivot = "PH"; // Pivot High
      } else if (centre.mid === minMid) {
        this.pivot = "PL"; // Pivot Low
      } else {
        this.pivot = null;
      }

      this.pivotBid = centre.bid;
      this.pivotAsk = centre.ask;
    }
  }
}
